mgmtsystem/virtualcenter.py

```diff
@@ -131,58 +127,142 @@ def __init__(self, hostname, username, password, **kwargs):
-    def api(self):
-        if not self._api:
-            self._api = _PsphereClient(self.hostname, self.username, self.password,
-                logger=self.logger)
-        return self._api
+    def service_instance(self):
+        """An instance of the service"""
+        if not self._service_instance:
+            self._service_instance = SmartConnect(host=self.hostname, user=self.username,
+                                                  pwd=self.password)
+        return self._service_instance
+
+    @property
+    def content(self):
+        """The content node"""
+        if not self._content:
+            self._content = self.service_instance.RetrieveContent()
+        return self._content
 
     @property
     def version(self):
-        return LooseVersion(self.api.si.content.about.version)
+        """The product version"""
+        return LooseVersion(self.content.about.version)
 
     @property
     def default_resource_pool(self):
         return self.kwargs.get("default_resource_pool", None)
 
-    def _get_vm(self, vm_name, force=False):
-        """ Returns a vm from the VI object.
+    def _get_obj_list(self, vimtype, folder=None):
+        """Get a list of objects of type ``vimtype``"""
+        if not folder:
+            folder = self.content.rootFolder
+        container = self.content.viewManager.CreateContainerView(folder, [vimtype], True)
+        return container.view
+
+    def _get_obj(self, vimtype, name, folder=None):
+        """Get an object of type ``vimtype`` with name ``name`` from Vsphere"""
+        obj = None
+        for item in self._get_obj_list(vimtype, folder):
+            if item.name == name:
+                obj = item
+                break
+        return obj
+
+    def _build_filter_spec(self, begin_entity, property_spec):
+        """Build a search spec for full inventory traversal, adapted from psphere"""
+        # Create selection specs
+        selection_specs = [vmodl.query.PropertyCollector.SelectionSpec(name=ss)
+                           for ss in SELECTION_SPECS]
+        # Create traversal specs
+        traversal_specs = []
+        for spec_values in TRAVERSAL_SPECS:
+            spec = vmodl.query.PropertyCollector.TraversalSpec()
+            spec.name = spec_values['name']
+            spec.type = spec_values['type']
+            spec.path = spec_values['path']
+            if spec_values.get('select_indices'):
+                spec.selectSet = [selection_specs[i] for i in spec_values['select_indices']]
+            traversal_specs.append(spec)
+        # Create an object spec
+        obj_spec = vmodl.query.PropertyCollector.ObjectSpec()
+        obj_spec.obj = begin_entity
+        obj_spec.selectSet = traversal_specs
+        # Create a filter spec
+        filter_spec = vmodl.query.PropertyCollector.FilterSpec()
+        filter_spec.propSet = [property_spec]
+        filter_spec.objectSet = [obj_spec]
+        return filter_spec
+
+    def _update_obj(self, obj):
```

Comment:
**REQUIRED** this method returns a updated copy, then method name **MUST** reflect that